File: src/main/java/wiresegal/psionup/common/PsionicUpgrades.java
package wiresegal.psionup.common;

import net.minecraft.block.Block;
import net.minecraft.item.Item;
import net.minecraft.util.SoundEvent;
import net.minecraftforge.client.event.ColorHandlerEvent;
import net.minecraftforge.event.RegistryEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.SidedProxy;
import net.minecraftforge.fml.common.event.*;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.registry.EntityEntry;
import net.minecraftforge.fml.relauncher.Side;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import wiresegal.psionup.common.block.ModBlocks;
import wiresegal.psionup.common.command.CommandPsiLearn;
import wiresegal.psionup.common.command.CommandPsiUnlearn;
import wiresegal.psionup.common.core.CommonProxy;
import wiresegal.psionup.common.core.PsionicSoundEvents;
import wiresegal.psionup.common.entity.ModEntities;
import wiresegal.psionup.common.items.ModItems;
import wiresegal.psionup.common.lib.LibMisc;

@Mod(modid = LibMisc.MOD_ID, name = LibMisc.MOD_NAME, version = LibMisc.VERSION, dependencies = LibMisc.DEPENDENCIES, acceptedMinecraftVersions = LibMisc.VERSIONS)
public class PsionicUpgrades {
	@Mod.Instance
	public static PsionicUpgrades INSTANCE = null;
	
	@SidedProxy(serverSide = LibMisc.PROXY_COMMON, clientSide = LibMisc.PROXY_CLIENT)
	public static CommonProxy PROXY;
	
	public static final Logger LOGGER = LogManager.getLogger(LibMisc.MOD_ID);
	
	@Mod.EventHandler
	public static void preinit(FMLPreInitializationEvent e) {
		PROXY.pre(e);
	}
	
	@Mod.EventHandler
	public static void init(FMLInitializationEvent e) {
		PROXY.init(e);
	}
	
	@Mod.EventHandler
	public static void postinit(FMLPostInitializationEvent e) {
		PROXY.post(e);
	}
	
	@Mod.EventHandler
	public static void serverStarting(FMLServerStartingEvent e) {
		e.registerServerCommand(new CommandPsiLearn());
		e.registerServerCommand(new CommandPsiUnlearn());
	}
	
	@Mod.EventBusSubscriber(modid = LibMisc.MOD_ID)
	public static class RegistryEvents {
		@SubscribeEvent
		public static void blocks(RegistryEvent.Register<Block> e) {
			ModBlocks.register(e.getRegistry());
		}
		
		@SubscribeEvent
		public static void items(RegistryEvent.Register<Item> e) {
			ModItems.register(e.getRegistry());
		}
		
		@SubscribeEvent
		public static void entities(RegistryEvent.Register<EntityEntry> e) {
			ModEntities.register(e.getRegistry());
		}
		
		@SubscribeEvent
		public static void sounds(RegistryEvent.Register<SoundEvent> e) {
			PsionicSoundEvents.registerSounds(e.getRegistry());
		}
	}
	
	@Mod.EventBusSubscriber(modid = LibMisc.MOD_ID, value = Side.CLIENT)
	public static class ClientEvents {
		@SubscribeEvent
		public static void blockColors(ColorHandlerEvent.Block e) {
			ModBlocks.registerBlockColors(e.getBlockColors());
		}
	}
}

File: src/main/java/wiresegal/psionup/common/PsionicUpgrades.kt
package wiresegal.psionup.common

import net.minecraft.launchwrapper.Launch
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.common.SidedProxy
import net.minecraftforge.fml.common.event.FMLInitializationEvent
import net.minecraftforge.fml.common.event.FMLPostInitializationEvent
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent
import net.minecraftforge.fml.common.event.FMLServerStartingEvent
import org.apache.logging.log4j.LogManager
import wiresegal.psionup.common.command.CommandPsiLearn
import wiresegal.psionup.common.command.CommandPsiUnlearn
import wiresegal.psionup.common.core.CommonProxy
import wiresegal.psionup.common.lib.LibMisc

/**
 * @author WireSegal
 * Created at 8:29 AM on 3/20/16.
 */
@Mod(modid = LibMisc.MOD_ID, name = LibMisc.MOD_NAME, version = LibMisc.VERSION, dependencies = LibMisc.DEPENDENCIES)
class PsionicUpgrades {
    companion object {
        @Mod.Instance(LibMisc.MOD_ID)
        lateinit var instance: PsionicUpgrades

        @SidedProxy(serverSide = LibMisc.PROXY_COMMON,
                clientSide = LibMisc.PROXY_CLIENT)
        lateinit var proxy: CommonProxy

        val LOGGER = LogManager.getLogger(LibMisc.MOD_ID)

        val DEV_ENVIRONMENT: Boolean by lazy {
            Launch.blackboard["fml.deobfuscatedEnvironment"] as Boolean
        }
    }

    @Mod.EventHandler
    fun preInit(event: FMLPreInitializationEvent) {
        proxy.pre(event)
    }

    @Mod.EventHandler
    fun init(event: FMLInitializationEvent) {
        proxy.init(event)
    }

    @Mod.EventHandler
    fun post(event: FMLPostInitializationEvent) {
        proxy.post(event)
    }

    @Mod.EventHandler
    fun serverStartingEvent(e: FMLServerStartingEvent) {
        e.registerServerCommand(CommandPsiLearn());
        e.registerServerCommand(CommandPsiUnlearn());
    }
}
